[WIP] Nullable embedded objects.

lcobucci opened the pull request:
The idea was to have a simple (and clean) way to override the nullable property of embedded objects, so I've
added the `nullable` attribute on the `@Embedded` annotation and it have 3 possible values:
- NULL: The nullable option that was defined on the attributes of the embeddable class won't be overriden;
- TRUE: All attributes of the embeddable class will be marked as nullable and the embeddable instance only would be created when data is not NULL;
- ~~FALSE: All attributes of the embeddable class will be marked as non-nullable.~~

There's a lot of things to be improved (mostly on UnitOfWork), but it's fully working with basic tests as you can see on ValueObjectsTest::testCRUDOfNullableEmbedded() case.

I would appreciate a lot your opinions!

lcobucci committed:
Initial work of nullable embedded objects.

There's a lot of things to be improved (mostly on UnitOfWork), but it's fully working with basic tests.
The idea was to have a simple (and clean) way to override the nullable property of embedded objects, so I've
added the nullable attribute on the Embedded annotation and it have 3 possible values:

- NULL: The nullable option that was defined on the attributes of the embeddable class won't be overriden;
- TRUE: All attributes of the embeddable class will be marked as nullable;
- FALSE: All attributes of the embeddable class will be marked as non-nullable.

The usage is quite simple and can be seen at the ValueObjectsTest::testCRUDOfNullableEmbedded() test case.

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/ValueObjectsTest.php:
```diff
@@ -242,6 +243,48 @@ public function testThrowsExceptionOnInfiniteEmbeddableNesting($embeddableClassN
         ));
     }
 
+    public function testCRUDOfNullableEmbedded()
```
@lcobucci consider splitting this test into smaller tests depending on each other via `@depends`

lcobucci replied:
Perfect, I followed the other test case (testCRUD). But using ```@depends``` is way better.

lcobucci replied:
I'll also change the name of the classes that I've created on that file to match the issue on JIRA (DDC-3529)

lcobucci committed:
Renaming test classes to match with the JIRA issue.

lcobucci committed:
Splitting testCRUDOfNullableEmbedded() into small tests.

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -2768,6 +2770,34 @@ public function createEntity($className, array $data, &$hints = array())
+     * @param array $data
+     * @param ClassMetadata $class
+     */
+    private function removeNullableEmbeddedReferences(array &$data, ClassMetadata $class)
```
@Ocramius I've tried to keep the things simple here, but if we have a huge amount of data or a large result set I think we might have a little delay to process. I just don't have any idea to improve this =/

1ed replied:
What about something like this:

```php
public function createEntity($className, array $data, &$hints = array())
{
    // ...
    foreach ($data as $field => $value) {
        if (null === $value && $this->isNullableEmbeddedField($class, $field)) {
            continue;
        }

        if (isset($class->fieldMappings[$field])) {
            $class->reflFields[$field]->setValue($entity, $value);
        }
    }
    // ...
}

private function isNullableEmbeddedField(ClassMetadata $class, $field)
{
  return isset(
      $class->fieldMappings[$field]['declaredField'],
      $class->embeddedClasses[$name = $class->fieldMappings[$field]['declaredField']]
  ) ? true === $class->embeddedClasses[$name]['nullable'] : false;
}
```

lcobucci replied:
Thanks @1ed, its a good alternative. I'll play a lit bit now :smile:

lcobucci replied:
@Ocramius, @1ed and @beberlei what do you think to extract that logic from ```UnitOfWork``` into ```ClassMetadata```? It seems to be proper location, considering that ```UnitOfWork``` is already huge.

I thought on something like ```ClassMetadata::updateEntity($entity, array $data)``` or  ```ClassMetadata::applyData($entity, array $data)```.

1ed commented:
:+1: that would be great if it could fit in 2.5

lcobucci committed:
Moving the responsibility of using the column data or not to fill the entity to the class metadata.

lcobucci commented:
I have implemented @1ed suggestion with that minor refactor.
We should update the interface, but it seems better to me.

alsar commented:
@lcobucci What is the status of this? 
Because it's only one few days left until the feature-freeze for the 2.5 release. It would be great if this could make it into 2.5.

Ocramius commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -3310,4 +3316,29 @@ public function getSequencePrefix(AbstractPlatform $platform)
         return $sequencePrefix;
     }
+
+    public function applyData($entity, array $data)
```
This method naming is a bit weird: can you document it?

lcobucci replied:
I'll rename it. It is really weird.
What do you think about ```populateEntity```

Ocramius commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -3310,4 +3316,29 @@ public function getSequencePrefix(AbstractPlatform $platform)
+    public function applyData($entity, array $data)
+    {
+        foreach ($data as $field => $value) {
+            if ($this->shouldUseData($field, $value)) {
```
Is there a way to cache this method call?

lcobucci replied:
It can be done, but do you think is needed? I mean, the operations aren't too complex and they're based on attributes that are already caching the information...

lcobucci commented:
It is "working" @alsar, but we have to clean some stuff to make it better. I'll send some commits right now.

lcobucci committed:
Renaming and documenting methods.

lcobucci committed:
Missing test. When @Embedded(nullable=false)

stof commented:
> FALSE: All attributes of the embeddable class will be marked as non-nullable.

this case looks weird to me. It changes an optional field to a required one in the embedded object

lcobucci committed:
Merge remote-tracking branch 'upstream/master'

lcobucci committed:
Adding nullable property on test.

lcobucci commented:
@stof please give us more arguments, let's debate :smile:.

In 6340eece56c8088ddb343fb76258c3545e872b2a I added a test for that cenario and I believe that in the "country" context is the only way to make the embeddable required, but when dealing with complex value object it can be weird.

The main question that comes to me is: should we have really complex value objects?

deeky666 commented:
I tend to agree with @stof about the `FALSE` case. Why would one want to make an optional field of the VO required? This really changes the VO semantics and should probably be another VO instead. Would need a good use case IMO otherwise the `FALSE` option does not make much sense to me.

deeky666 commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -3226,8 +3227,9 @@ public function mapEmbedded(array $mapping)
+     * @param boolean $nullable
      */
-    public function inlineEmbeddable($property, ClassMetadataInfo $embeddable)
+    public function inlineEmbeddable($property, ClassMetadataInfo $embeddable, $nullable)
```
Is that extra parameters really required? Can't you use `$this->embeddedClasses[$property]['nullable']` in this method instead?

lcobucci replied:
You're right, my mistake.

lcobucci committed:
Removing unnecessary parameter.

lcobucci commented:
@deeky666 the feature that I need is the ```TRUE``` case. As I said before, with simple VO (like the one in ```ValueObjectsTest```) it can make sense but when dealing with complex value objects it is weird.

I can remove the ```FALSE``` case so we can merge this feature to doctrine 2.5, and the future will tell us if it was a wise choice.

lcobucci commented:
So, do we remove the ```FALSE``` case guys?

alsar commented:
@lcobucci lets remove `false` for now and get this merged into 2.5. 
I can then be added to 2.6 if it will be needed.

lcobucci committed:
Merge remote-tracking branch 'upstream/master'

lcobucci committed:
Don't allow marking optional embeddable fields as mandatory.

lcobucci commented:
I've removed the ```false``` (by ignoring it), should we raise an error when somene try to use false as value for the nullable option on embedded objects?

lcobucci commented:
@Ocramius @deeky666 do we need something else in order to merge this?

Ocramius commented:
> should we raise an error when somene try to use false as value for the nullable option on embedded objects?

Yes please

Ocramius commented:
Also: what happens when a nullable embeddable has null fields for properties explicitly mapped as non-null?

lcobucci commented:
> Also: what happens when a nullable embeddable has null fields for properties explicitly mapped as non-null?

The non-null fields would be mapped as nullable.

Ocramius commented:
Which in turn allows for invalid state to be stored/persisted in the embeddable. I think that's quite the deal breaker.

lcobucci commented:
> Which in turn allows for invalid state to be stored/persisted in the embeddable. I think that's quite the deal breaker.

Ok, but the embeddable should be responsible for dealing with that by not accepting the creation of the object without the required fields, no black magic here.

There's no way to make the embedded optional in one place and required in another without overriding the embeddable fields configuration.

itised commented:
Would it be possible to instead let @AttributeOverides apply to embedded objects? That would allow more control over marking individual fields as nullable, but also allow much more control over other properties, and using a concept that is already part of doctrine.

Ocramius commented:
I think the only way to make embeddables nullable (in a safe way) right now is to allow nullable embeddables only when all `@Embeddable` fields are `nullable=true`.

lcobucci commented:
> Would it be possible to instead let @AttributeOverrides apply to embedded objects?

As of @schmittjoh said on the [draft of this PR](https://gist.github.com/lcobucci/9e68501cfad6931119d7) we already have a way to override embeddable mapping.

> I think the only way to make embeddables nullable (in a safe way) right now is to allow nullable embeddables only when all ```@Embeddable``` fields are nullable=true.

The point that I'm trying to explain here is that the safe way should be granted by the constructor not by the embeddable mapping. With the ```@AttributeOverrides``` we can can change (in a messy way) the mapping to make everything nullable and the instance will be created (with NULL on every single attribute).

On the draft (and on the test case) we have examples for what I'm saying and you can see that we won't have an invalid state.

Please @Ocramius help me to open my eyes and understand the problem you're saying.

Ocramius commented:
@Icobucci you have to remember that the ORM keeps consistency both at object and at DB level at heart. If an embeddable is nullable, then all of its fields should be nullable at DB level as well, leading to inconsistent state, at which point I'd simply say:

 - use a json-serialized manually-coded embeddable, knowing what you are facing
 - use a one-to-one association instead of an embeddable

lcobucci commented:
@Ocramius I understand that, but I disagree that the requirement of the embeddable should be the same on every context.

Looking at the DB level, (IMO) is not an inconsistent state to have the embeddable fields nullable on a table a required on another, because at the DB level what is important is the tables and by looking to them there is no reason to know if the field was declared by an entity or an embeddable.

Embeddables are a reusable set of values (we can also link with value objects of DDD, but I'll pass this argument for now) and if I want to say that, in a given context, I want to make that set optional I shouldn't be blamed.

By setting the ```nullable``` as ```TRUE``` on the ```@Embedded``` I'm saying that I'm well aware of what I'm doing and what will be the consequences.

I don't see embeddables strong enough to become a one-to-one association and weak enough to be mapped as a json-serialized object.

Ocramius commented:
> By setting the `nullable` as `TRUE` on the `@Embedded` I'm saying that I'm well aware of what I'm doing and what will be the consequences.

No, by setting `nullable` as `TRUE` you are making it `nullable`: you are not accepting the idea of invalid data being hydrated

"nullable" and "inconsistent" are very different concepts, and SQL DBMSs without struct-like fields aren't able to deal with this sort of data-integrity problem in a decent way, therefore it's simply better to avoid the problem in first place by putting a strict limitation instead of building more foot-gunning tools

lcobucci commented:
Well, we agree to disagree :smile:

We still have ways to bypass the original mapping with the ```@AttributeOverrides``` an that will lead us to an useless instance, which is pretty inconsistent to me.

You have arguments against and I have arguments in favour, but all of them was already written and we will be repeating and throwing words in the wind if we continue the debate.

If the you and the other doctrine maintainers don't think this is a good increment we should close this PR, agreed?

lcobucci commented:
@Ocramius I just saw your sample code, let me explain my POV (for one last time)...

```php
class Credentials // embeddable
{
    /** @var string (never null) */
    public $username;
    /** @var string (never null) */
    public $password;

    private function __construct($username, $password)
    {
        $this->username = $username;
        $this->password = $password;
    }
}

class User
{
    public $credentials; // embed here, required
}

class Client
{
    public $credentials; // embed here, nullable
}
```

The ORM will NEVER run:

```sql
INSERT INTO users (credentials_username, credentials_password) VALUES ('ocramius', null);
-- OR
INSERT INTO users (credentials_username, credentials_password) VALUES (null, 'hello');
-- OR
INSERT INTO users (credentials_username, credentials_password) VALUES (null, null);
-- OR
INSERT INTO client (credentials_username, credentials_password) VALUES ('ocramius', null);
-- OR
INSERT INTO client (credentials_username, credentials_password) VALUES (null, 'hello');
```

The only valid options are (at the ORM level, of course):

```sql
INSERT INTO users (credentials_username, credentials_password) VALUES ('ocramius', 'hello');
-- AND
INSERT INTO client (credentials_username, credentials_password) VALUES (null, null);
-- AND
INSERT INTO client (credentials_username, credentials_password) VALUES ('ocramius', 'hello');
```

One can execute the invalid queries at the DB level, and that will indeed lead us to an invalid state but that can cause trouble on other ORM areas as well (like the second level cache).

I understand what you're explaining, I just think that this issue can be reached with or without this PR.

Well, thanks for the discussion :smile:

Ocramius commented:
> I just think that this issue can be reached with or without this PR

That's exactly where I wanted you to go to: this starts to treat the ORM as a gateway to a SQL DBMS used as a NoSQL DBMS :-)

This sort of constraint must be at DB level as well, as the ORM must be able to operate with datasets that are not managed by it.

dirkluijk commented:
I really would like to have such thing.

My use case: an ``Order`` with two embedded ``Address``: ``billingAddress`` and ``invoiceAddress``. The first one is required, the second one is optional.

It seems I have to store the adresses in a separate table and make two associations, but I wouldn't like that because limitations of the current backoffice.

pinkeen commented:
Embeddables are very useful for value objects. Usually a value object can't have empty data and this is enforced in the constructor. When you assign `null` to a field with embedded object then all of it's columns will become null (if nullable). Then after hydration you will get a VO with empty data, which isn't a permitted state - your domain is fubared.

The only solutions I see:
- don't use embeddables for optional properties
- implement a method like isEmpty on each value object which really breaks a lot of stuff (ex. serialization) and is really ugly; you also have to remember not to do any checks like if `(null !== $this->optionalPropertyVO)`

alanbem commented:
Is this idea scrapped for good?

I had never dived into Doctrine internals so I've got the question: Is it impossible to check if embedded object is nullable and - during hydration - if all its fields in data-row are null.... than not hydrate it?  

I understand @Ocramius concerns but not all of us treat mapped objects like stupid data bags that easily go into invalid state. My model is mostly bulletproof in context of consistency and what I need is data mapper which is flexible enough to accommodate to my needs.
Unfortunately Doctrine often stands in the way of truly persistent agnostic model.

Ocramius commented:
>  Is it impossible to check if embedded object is nullable and - during hydration - if all its fields in data-row are null.... than not hydrate it? 

Yes, it is possible, but it requires some effort, and the performance overhead must be measured.

 > My model is mostly bulletproof in context of consistency and what I need is data mapper which is flexible enough to accommodate to my needs

Then write a custom mapper: doctrine is not a *flexible data mapper*, it is a data mapper with a very opinionated approach. Custom use-cases shouldn't pollute the core API with hacks that actually lead to more bugs than what they solve.

 > Unfortunately Doctrine often stands in the way of truly persistent agnostic model.

That is true, but the only solution to that is custom persisters and hydrators, and not necessarily ORM-based. We still didn't get around to design those.

acasademont commented:
Hi @Ocramius do you think this should be reopened as another bug in JIRA? I still think that this should be done somehow. Maybe a metadata column so that the underlyining embedadble properties have not to be set as nullable?

vvh-empora commented:
so what is the current state? this is a real missing feature!?

pinkeen commented:
I would love this feature!

We have a big project with lots of value object. Right now our code is polluted with `SomeVOInstance->isReallyEmpty()` code.

lcobucci commented:
Guys I think its pointless to discuss this without suggesting a solution to our problem.

When I wrote this PR I thought only about OOP and then @Ocramius said that people could have invalid state by inserting or updating the tables using SQL. I was like: who in f****** hell would be insane enough to insert wrong data into his own database? So I truly understand the frustration, but we can't say that our code is bullet-proof and allow possible issues - or maybe vulnerabilities - be injected into our dependencies.

I think that the only way to have this feature is to create it as an extension. Who "threat mapped objects like stupid data bags that easily go into invalid state" wouldn't use this modified hidrator/metadatafactory and everybody would be happy.

pinkeen commented:
I would be willing to (at least try) to implement this feature as an extension. Is there currently an elegant way to inject custom Hydrator/MetadataFactory (as defaults)? Could anybody point me in the right direction?

Ocramius commented:
This is a chicken-egg problem, folks. Not going to happen unless somebody finds a solution to the data consistency problem first.

Ocramius commented:
Also: using the `object` type is perfectly legit.

schmittjoh commented:
I think it might be best to use an extra column that has a 1 or 0 to check whether an embedded property should be null or not.

pinkeen commented:
@schmittjoh How do I do it transparently in the userland? Could you give me a hint?

schmittjoh commented:
Something along these lines should work:

```php
class Foo
{
     private $embeddable;
     private $isEmbeddableSet;

     public function getEmbeddable()
     {
         if ( ! $this->isEmbeddableSet) { return null; }

         return $this->embeddable;
     }
}

lcobucci commented:
We actually don't need a modified Hydrator as you can see in the files I've changed.

The MetadaFactory is easily injected on EntityManager configuration as you can see here: https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/EntityManager.php#L147

You can also create an extension of the AnnotationDriver and pass it to your configuration too.

pinkeen commented:
@lcobucci thanks, I will look into this.

alsar commented:
Does somebody know how they resolve this issue in Hibernate?

schmittjoh commented:
@pinkeen, it should be possible to hide the implementation details behind a Doctrine API, so that you do not need to deal with this in user-land, someone would need to create a pull-request though to allow for further discussion

@alsar, last time I checked, it was what I outlined above

vvh-empora commented:
hmmm... very tricky :| is there any known existing extension for this problem?

fabiocarneiro commented:
I don't understand the argument that even with someone manually modifying database the consistency of application should still be valid. 

To make an analog, you could in your domain specify that the only allowed values of an object are 1, 2 and 3, and then someone goes to database and update the value to 4. It would still produce invalid state, and could still have catastrophic consequences.

As @lcobucci said, it would be perfect fine to define it nullable in one place and not nullable in other places, and if we eventually follow the idea of having this as a separate table, it would still need to follow the same schema, probably meaning that all columns would be nullable or following the mapping of the embeddable. This basically means that the only way to store embeddables in relational databases is by using columns.

I personally never seen a reason to create a VO with null properties since it doesn't represent any value and the idea is in the opposite direction.

fabiocarneiro commented:
I've been testing this branch and I noticed that it doesn't work with multiple levels of embeddables. It creates the object no matter if all columns of all nested embeddables are null.

fprochazka commented:
What about allowing nullable embedables only for DBs that have [constraints](http://www.postgresql.org/docs/8.1/static/ddl-constraints.html)?

Ocramius commented:
Nope.
On Feb 23, 2016 13:07, "Filip Procházka" <notifications@github.com> wrote:

> What about allowing nullable embedables only for DBs that have constraints
> <http://www.postgresql.org/docs/8.1/static/ddl-constraints.html>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/1275#issuecomment-187674516>.
>

lcobucci commented:
@FabioBatSilva I didn't tested that kind of thing but I don't think would be something difficult to achieve. 

I also believe that @Ocramius's statement was good enough to end this (as a standard feature):

> That's exactly where I wanted you to go to: this starts to treat the ORM as a gateway to a SQL DBMS used as a NoSQL DBMS :-)
> This sort of constraint must be at DB level as well, as the ORM must be able to operate with datasets that are not managed by it.

About @fprochazka's comment:

We cannot expect to push a lot of logic to doctrine's shoulders (it already have too much to carry now). I believe that we need to find a way to solve #4568 and also that we should treat this feature as an extension (if it would be really needed after solving that issue).

fabiocarneiro commented:
@lcobucci I think you are accepting this argument too fast. Allowing mappings to happen doesn't mean that this will make a SQL DBMS to be used as DBMS. By doing this you don't start storing documents or objects. Its still tables and rows.

The thing is that the mapper as we have now, does not support proper optional objects. The goal of the mapper as I understand is to map any design you can possibly have in your code to relational databases, and having optional (null) properties is far too common to be ignored. Its the same as when we didn't have the embeddables feature and we would choose to design everything with scalar types just to fit the mapper.

lcobucci commented:
Oops you're correct @fabiocarneiro sorry hehehe.

No, I'm not accepting it too fast (it was a long discussion already :smile:).

With #4568 we will be able to use the `@AttributeOverrides` annotation as @schmittjoh commented on the gist that I created [before sending this PR](https://gist.github.com/lcobucci/9e68501cfad6931119d7). It may lead to the same problem that @Ocramius mentioned however it will not require any other change on doctrine - and only users that are 100% sure that the database is not handled by other thing rather than doctrine would use it.

fabiocarneiro commented:
@lcobucci as far as I understood, we could use attribute overrides to force all mappings of embeddables to be null, meaning that it could be different configuration in different columns. But how does that prevent the hydration of the object that has both columns nullable?

For my value objects, null are inconsistent state, but for the entity that has it as property its not. What I really want is the object not to be created rather than its properties having null.

`ORM must be able to operate with datasets that are not managed by it`. How exactly by adding this feature we would prevent this to happen?

lcobucci commented:
But @fabiocarneiro, that's exactly what's being stated on #4568.
In this PR I've fixed it partially and added the `nullable` attribute, we need to cover it properly and think about all edge cases and we don't need the `nullable` modifier because we can achieve it in different ways.

I think we should see with @Ocramius || @guilhermeblanco || @schmittjoh what we could do about that (but on the scope of that issue and not on this PR).

backbone87 commented:
There are database features that allow these kind checks in various ways at db level. Just because doctrine can not infer a schema that facilitates such features, doesnt mean it should assume these features are not available. also the mapping to schema conversion is just a secondary feature of doctrine to ease database creation and not to replace each and every (vendor-dependant) db feature.

i would really like this feature to be added.

the eal problem comes when all fields in the embeddable are nullable and the embeddable itself is nullable, because we can not distinquish between a null embeddable and an embeddable containing only null values. a separate boolean-like column is required in this case.

Ocramius commented:
> the eal problem comes when all fields in the embeddable are nullable and the embeddable itself is nullable, because we can not distinquish between a null embeddable and an embeddable containing only null values. a separate boolean-like column is required in this case.

That is a sensible solution, although a joined one-to-one is simpler and better for this scenario (and adding a 1-to-1 join is a low cost operation, if indexed correctly)

Padam87 commented:
Sorry to up this, I read back a lot, but it is possible I have missed something. What about this?:

If all fields are null, check the embeddable configuration. If the configuration specifies that it is nullable, set is as `null`. If it is not nullable, create the object in **every** case, even if all fields are null.

```php
$address = new Address(); // all fields nullable, and null
$order->setShippingAddress($address);

$em->persist($order);
$em->flush();
$em->refresh($order);

// null (assuming the configuration specifies nullable=true)
$order->getShippingAddress()

// instanceof Address, empty (assuming the configuration specifies nullable=null)
$order->getShippingAddress()
```

enumag commented:
@Padam87 pretty much wrote down my own thought about this. Can we get a response to his comment? cc @Ocramius @lcobucci

Ocramius commented:
It is a possible implementation, but nullability on all embeddable fields would then be disallowed (need at least one non-null field).

That's a good idea for 3.x, so if anybody wants to help out, start from `develop` :-)
